src/client/WebSocketClient.ts

```diff
@@ -502,19 +521,117 @@ export class WebSocketClient extends EventEmitter {
+    }
+
+    // Only regenerate if a change was made
+    if (updateSubscriptionCache) {
+      this.generateSubscriptionsCache();
+    }
+
+    this._subscribe(channel);
+  }
+
+  private _subscribe(channel: WebSocketChannel | WebSocketChannel[]): void {
     this.sendMessage({
       channels: Array.isArray(channel) ? channel : [channel],
       type: WebSocketRequestType.SUBSCRIBE,
     }).finally(() => {});
   }
 
   unsubscribe(channel: WebSocketChannelName | WebSocketChannel | WebSocketChannel[]): void {
+    let updateSubscriptionCache = false;
+    if (typeof channel === 'string' && this._subscriptions[channel]) {
+      // Unsubscribing for all products
+      delete this._subscriptions[channel];
+
+      updateSubscriptionCache = true;
+    } else {
+      const channels = this.mapChannels(channel);
+      for (const chan of channels) {
+        if (this._subscriptions[chan.name]) {
+          if (undefined !== chan.product_ids) {
+            // Remove product ids from the channel
+            this._subscriptions[chan.name].product_ids = this._subscriptions[chan.name].product_ids.filter(
+              x => !chan.product_ids?.includes(x)
+            );
+
+            // If no more products subscribed on the channel, delete its reference
+            if (0 === this._subscriptions[chan.name].product_ids.length) {
+              delete this._subscriptions[chan.name];
+            }
+          } else {
+            // A channel with just the name and no product ids is equivalent to unsubscribe from all of that channel
+            delete this._subscriptions[chan.name];
+          }
+
+          updateSubscriptionCache = true;
+        }
+      }
+    }
+
+    // Only regenerate if a change was made
+    if (updateSubscriptionCache) {
+      this.generateSubscriptionsCache();
+    }
+
+    this._unsubscribe(channel);
+  }
+
+  private _unsubscribe(channel: WebSocketChannelName | WebSocketChannel | WebSocketChannel[]): void {
     this.sendMessage({
       channels: this.mapChannels(channel),
       type: WebSocketRequestType.UNSUBSCRIBE,
     }).finally(() => {});
   }
 
+  private generateSubscriptionsCache(): void {
+    const channels: WebSocketChannel[] = [];
+    for (const channelName in this._subscriptions) {
+      if (Object.prototype.hasOwnProperty.call(this._subscriptions, channelName)) {
+        const chan = this._subscriptions[channelName];
+
+        // Sort product IDs so they are returned in a predictable order
+        chan.product_ids.sort();
+
+        channels.push(chan);
+      }
+    }
+
+    // Sort channels so they are returned in a predictable order
+    this._subscriptions_cache = channels.sort((a, b): number => {
+      if (a.name > b.name) {
+        return 1;
+      }
+      return -1; // assume a's name is less than b's name
+    });
+  }
+
+  /**
+   * Returns all the channels and products that have been subscribed.
+   */
+  get subscriptions(): WebSocketChannel[] {
+    return this._subscriptions_cache;
```

Comment:
This cache array should be kept up-to-date as changes happen, so we should just need to return it.